### ISSUE TITLE
Unexpected 'Fragment cannot be spread' error

### ISSUE THREAD

**Issue by AlistairB** [morpheusgraphql/morpheus-graphql#493]
Hi,

I am trying to use interfaces in my query using morpheus client. My gql file looks something like (excluded some excessive detail. Let me know if you want my full file):

```graphql
"""
An object with an ID.
"""
interface Node {
  """
  ID of the object.
  """
  id: ID!
}

"""
Represents a Git object.
"""
interface GitObject {
  id: ID!
}

"""
Represents a Git reference.
"""
type Ref implements Node {
  id: ID!

  """
  The object the ref points to.
  """
  target: GitObject!
}

type Commit implements GitObject & Node {

  id: ID!

  """
  Authorship details of the commit.
  """
  author: String
}

type Tree implements GitObject & Node {
  """
  A list of tree entries.
  """
  treeName: String

  id: ID!
}
```

And my query looks something like:

```haskell
defineByDocumentFile
  "../resource/minimal-github.graphql"
  [gql|
    query FetchRepoHeadCommit ($repoOwner: String!, $repoName: String!) {
      repository(name: $repoName, owner: $repoOwner) {
        defaultBranchRef {
          target {
            ... on Commit {
               author
            }
          }
        }
      }
    }
  |]
```

This fails with `Fragment cannot be spread here as objects of type \"GitObject\" can never be of type \"Commit\`. Changing `  target: GitObject!` to `target: Commit!` makes it work. This confuses me as I have `type Commit implements GitObject`.

My expectation is that this should work based on `2` from https://github.com/morpheusgraphql/morpheus-graphql#interface . Am I missing something?

**Comment by nalchevanidze** [comment]
Thanks, you are right.

Query validation was written bevore we supported "interfaces", so i forgot to update it too.

I will fix it

**Comment by nalchevanidze** [comment]
@AlistairB i started fixin it in  #495

**Comment by nalchevanidze** [comment]
@AlistairB please try it with new master

**Comment by AlistairB** [comment]
Unfortunately I am still getting the same error building with morpheus based on commit https://github.com/morpheusgraphql/morpheus-graphql/commit/57de5110e95b834fb92aba393757923273241328 (current head on master).

**Comment by nalchevanidze** [comment]
ooh. that weird i wrote tests for it. can you share full schema?

**Comment by nalchevanidze** [comment]
and query with fragment

**Comment by AlistairB** [comment]
If you clone https://github.com/AlistairB/morpheus-repro and `stack build` it should reproduce the issue. It will only compile if you change https://github.com/AlistairB/morpheus-repro/blob/master/minimal-github.graphql#L89 to `target: Commit!`.

**Comment by nalchevanidze** [comment]
so, if we have schema.

```gql
interface GitObject {
  id: ID!
}

type Commit implements GitObject & Node {
  id: ID!
  author: String
}

type Query  {
  target: GitObject!
}
```

your query would look like:

```gql
{
   target {
    ... on Commit {
         author
         }
   }
}
```

and would be invalid. as long `target` guarantees to provide fields defined by `GitObject` nut by `Commit`.

if we had



so, if we have schema where target has concrete type `Commit`.

```gql
interface GitObject {
  id: ID!
}

type Commit implements GitObject & Node {
  id: ID!
  author: String
}

type Query  {
  target: Commit!
}
```

we can query on it `GitObject` and `Commit`. 

```gql
{
   t1: target {
    ... on Commit {
         author
         }
   }
   t2: target {
    ... on GitObject {
         author
         }
   }
}
```

**Comment by AlistairB** [comment]
Hmm, I'm not sure if I understand. My understanding is that you can use the spread operator to conditionally get data IF the type matches.

For example using https://developer.github.com/v4/explorer/

```graphql
query { 
  repository(owner: "facebook", name: "react") {
    defaultBranchRef {
      target {
        id
        ... on Commit {
          author {
            name
          }
        }
      }
    }
  }
}
```
will fetch all this data, as `target` is in fact a `Commit`.

On the other hand:

```
query { 
  repository(owner: "facebook", name: "react") {
    defaultBranchRef {
      target {
        id
        ... on Blob {
          byteSize
        }
      }
    }
  }
}
```
Will only fetch the id, but will not return an error. This is because `target: GitObject` and `type Blob implements GitObject`, so it could be a `Blob`.

And you can combine them:

```graphql
query { 
  repository(owner: "facebook", name: "react") {
    defaultBranchRef {
      target {
        id
        ... on Commit {
          author {
            name
          }
        }
        ... on Blob {
          byteSize
        }        
      }
    }
  }
}
```

In terms of the schema I am using, I have directly copied the bits I need from the official github public schema. See https://docs.github.com/en/graphql/overview/public-schema .

This query with the schema does work on github which makes me think it is valid?

**Comment by nalchevanidze** [comment]
okey.  may i got something wrong.  what you say looks like type casting. 
i will read specifications and fix it.

**Comment by nalchevanidze** [comment]
@AlistairB can you check again , i fixed it #498

**Comment by nalchevanidze** [comment]
@AlistairB validation workded well bot there is some problem with type Generation that i am fixing in #499

**Comment by nalchevanidze** [comment]
@AlistairB now it should finally work

**Comment by AlistairB** [comment]
Thanks David! This is now working for most cases that I tested. The below is happy and works.

```graphql
    query FetchRepoHeadCommit ($repoOwner: String!, $repoName: String!) {
      repository(name: $repoName, owner: $repoOwner) {
        defaultBranchRef {
          target {
            id
            ... on Commit {
              committedDate
            }
          }
        }
      }
    }
```

However, I get an error if I remove returning the `id`. eg.

```graphql
    query FetchRepoHeadCommit ($repoOwner: String!, $repoName: String!) {
      repository(name: $repoName, owner: $repoOwner) {
        defaultBranchRef {
          target {
            ... on Commit {
              committedDate
            }
          }
        }
      }
    }
```

produces `No Empty fields on \"RepositoryDefaultBranchRefTargetGitObject\"`. It seems that you must fetch something from the interface otherwise it errors. I have pushed a new commit demonstrating this failure to https://github.com/AlistairB/morpheus-repro

Possibly worth noting that I believe putting interface level fields inside the spread is also valid. eg.

```graphql
    query FetchRepoHeadCommit ($repoOwner: String!, $repoName: String!) {
      repository(name: $repoName, owner: $repoOwner) {
        defaultBranchRef {
          target {
            ... on Commit {
              id
              committedDate
            }
          }
        }
      }
    }
```

I think this is working aside from the above error. Although, I don't think supporting this style is super important.

**Comment by nalchevanidze** [comment]
thanks @AlistairB. 

i fixed it in #500

**Comment by AlistairB** [comment]
Awesome, now working! thanks!
